### PR TITLE
feat: support import attributes

### PR DESCRIPTION
**Pull request by chentsulin**
### What?

Add support to import attributes.

### Why?

The old import assertions syntax is deprecated and the proposal is updated.  See https://github.com/babel/babel/pull/15536

### How?

Add support to import attributes and keep old import assertions working by using `@babel/plugin-syntax-import-attributes` with the `deprecatedAssertSyntax` option.

Docs: https://babeljs.io/docs/babel-plugin-syntax-import-attributes

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: ff58490602301437c18cd55a2af0d5069df23258

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: ff58490602301437c18cd55a2af0d5069df23258

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: ff58490602301437c18cd55a2af0d5069df23258

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Review by huozhi — COMMENTED** [review]
Thanks for the PR,  the integration test `test/integration/import-assertion` also needs to be updated. For swc options you  also need to enable `keepImportAttributes` and drop `keepImportAssertions` flag

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| buildDuration | 10.7s | 11s | ⚠️ +285ms |
| buildDurationCached | 6s | 6.1s | ⚠️ +110ms |
| nodeModulesSize | 200 MB | 200 MB | ⚠️ +732 B |
| nextStartRea..uration (ms) | 421ms | 422ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| 170-HASH.js gzip | 26.9 kB | 26.9 kB | N/A |
| 199.HASH.js gzip | 181 B | 185 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 239 B | 241 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.5 kB | 98.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| _app-HASH.js gzip | 195 B | 194 B | N/A |
| _error-HASH.js gzip | 183 B | 182 B | N/A |
| amp-HASH.js gzip | 501 B | 501 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 255 B | ✓ |
| head-HASH.js gzip | 349 B | 350 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.6 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 309 B | N/A |
| script-HASH.js gzip | 384 B | 384 B | ✓ |
| withRouter-HASH.js gzip | 307 B | 306 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 521 B | N/A |
| Overall change | 541 B | 541 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.8 kB | 93.8 kB | N/A |
| page.js gzip | 146 kB | 146 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 627 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.4 kB | 37.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | chentsulin/next.js feat/support-import-attributes | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 168 kB | 168 kB | ✓ |
| app-page-exp..prod.js gzip | 94 kB | 94 kB | ✓ |
| app-page-tur..prod.js gzip | 94.7 kB | 94.7 kB | ✓ |
| app-page-tur..prod.js gzip | 89.3 kB | 89.3 kB | ✓ |
| app-page.run...dev.js gzip | 138 kB | 138 kB | ✓ |
| app-page.run..prod.js gzip | 88.6 kB | 88.6 kB | ✓ |
| app-route-ex...dev.js gzip | 24 kB | 24 kB | ✓ |
| app-route-ex..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| app-route.ru...dev.js gzip | 23.4 kB | 23.4 kB | ✓ |
| app-route.ru..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.4 kB | 49.4 kB | ✓ |
| Overall change | 930 kB | 930 kB | ✓ |

</details>

</details>
Commit: 064a60c2888a910551530d6718e8332ddbb31566

**Comment by huozhi** [comment]
> @huozhi I can't find any keepImportAssertions in the current codebase and keepImportAttributes has already been enabled:
@chentsulin nice, looks like it's already updated, all good!

For prettier errors we could use prettier ignore comments, add a TODO comment and upgrade prettier later in separate PR

**Comment by chentsulin** [comment]
@huozhi Tried. The approach didn't work because it's a parsing error:

```
✖ prettier found some errors. Please fix them and try committing again.


[error] test/integration/import-attributes/pages/es.js: SyntaxError: Unexpected token, expected "(" (3:33)
[error]   1 | // TODO: upgrade prettier to 3.1.1 to support import attributes
[error]   2 | // prettier-ignore
[error] > 3 | import data from '../data' with { type: 'json' }
[error]     |                                 ^
[error]   4 |
[error]   5 | export default function Es() {
[error]   6 |   return data.foo
[error] test/integration/import-attributes/pages/ts.ts: SyntaxError: ';' expected. (3:28)
[error]   1 | // TODO: upgrade prettier to 3.1.1 to support import attributes
[error]   2 | // prettier-ignore
[error] > 3 | import data from '../data' with { type: 'json' }
[error]     |                            ^
[error]   4 |
[error]   5 | export default function Ts() {
[error]   6 |   return data.foo
 ELIFECYCLE  Command failed with exit code 1.
 ```
 
 So I added the files to the `.prettierignore` file instead.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: b8d9597c1621db5c1908b3fefc6e9248a4658a44

`pnpm test-dev test/e2e/app-dir/not-found/basic/index.test.ts` 
- app dir - not-found - basic > with runtime = edge > should render the 404 page when the file is removed, and restore the page when re-added

<details>
<summary>Expand output</summary>

  ● app dir - not-found - basic › with runtime = edge › should render the 404 page when the file is removed, and restore the page when re-added

    TIMED OUT: Root Not Found

    My page

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (e2e/app-dir/not-found/basic/index.test.ts:106:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/integration/import-attributes/pages/ts.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import data from '../data' with { type: 'json' }
```
Comment (huozhi) [inline test/integration/import-attributes/pages/ts.ts:1]: eslint is failing
```
/root/actions-runner/_work/next.js/next.js/test/integration/import-attributes/pages/ts.ts
  1:27  error  Parsing error: ';' expected

✖ 1 problem (1 error, 0 warnings)
```

Comment (chentsulin) [reply test/integration/import-attributes/pages/ts.ts:1]: I created a separate PR to upgrade typescript-eslint: https://github.com/vercel/next.js/pull/59514

Comment (chentsulin) [reply test/integration/import-attributes/pages/ts.ts:1]: @huozhi Should I bump TS to 5.3.3 in this PR too? I'm not sure but it may fix it.

Comment (styfle) [reply test/integration/import-attributes/pages/ts.ts:1]: Let's bump TS in a separate PR

Comment (chentsulin) [reply test/integration/import-attributes/pages/ts.ts:1]: I can do rebase after https://github.com/vercel/next.js/pull/59494 get merged.

Comment (chentsulin) [reply test/integration/import-attributes/pages/ts.ts:1]: No, it still can't parse the syntax correctly. I believe we need TypeScript 5.3 to solve the issue.

```
/root/actions-runner/_work/next.js/next.js/test/integration/import-attributes/pages/ts.ts
  1:27  error  Parsing error: ';' expected

✖ 1 problem (1 error, 0 warnings)
```

Comment (huozhi) [reply test/integration/import-attributes/pages/ts.ts:1]: @chentsulin We've upgraded typescript to 5.3 on canary, you can update the branch now 🙏

Comment (chentsulin) [reply test/integration/import-attributes/pages/ts.ts:1]: The branch has been updated. 🙏

Comment (styfle) [reply test/integration/import-attributes/pages/ts.ts:1]: Looks like there are still merge conflicts in pnpm-lock.yaml

Maybe you can try regenerating it from the main branch

**Review by styfle — APPROVED** [review]

**Comment by kwonoj** [comment]
@chentsulin appreciate for the effort, but for now we reverted this change in https://github.com/vercel/next.js/pull/65001.

While debugging test failures, found out newly added test's assertion is too wide to misdetermine failing tests in turbopack as passing (https://github.com/vercel/next.js/pull/65000), which means this change does not work correctly on the turbopack.

**Pull request merged**